File: code/backend/java/core/src/main/java/games/paths/core/service/match/TimeAdvancementService.java
package games.paths.core.service.match;

import games.paths.core.model.match.MatchStatuses;
import games.paths.core.model.match.TurnStatuses;
import games.paths.core.model.match.event.TimeAdvanced;
import games.paths.core.port.event.DomainEventPublisher;
import games.paths.core.port.match.EventExecutionPort.EdgeStateOutcome;
import games.paths.core.port.match.LocationEntryPort;
import games.paths.core.port.match.TimeAdvancementPort;
import games.paths.core.port.match.TurnCyclePort.TurnCycleException;
import games.paths.core.port.match.TurnCycleStorePort;
import games.paths.core.port.match.TurnCycleStorePort.CharacterTurnView;
import games.paths.core.port.match.TurnCycleStorePort.ClockLabels;
import games.paths.core.port.match.TurnCycleStorePort.MatchView;
import games.paths.core.port.match.TurnCycleStorePort.QueueRow;
import games.paths.core.port.match.UserAccessPort;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * TimeAdvancementService - time advancement & clock cycle engine (Step 25).
 *
 * <p>Adds a voluntary sleep action and a clock query, and advances the match
 * clock when every active character is "done" (sleeping or out of energy — in
 * single-player the list is size 1). On time-end it logs the advance, wakes the
 * characters, rebuilds {@code gaming_turn_queue} reusing the Step 24
 * {@link TurnPriorityCalculator}, and publishes a {@link TimeAdvanced} event.</p>
 *
 * <p>See {@code documentation_v0/Step25_TimeAdvancementClockCycle.md}.</p>
 */
public class TimeAdvancementService implements TimeAdvancementPort {

    private static final String DEFAULT_LANG = "en";

    private final TurnCycleStorePort store;
    private final UserAccessPort userAccessPort;
    private final DomainEventPublisher eventPublisher;
    private final TimeStartRecoveryService recoveryService;
    private final WeatherSelectionService weatherService;
    /**
     * Step 33 — the engine that runs the automatic events a time-start collected.
     *
     * <p>Injected through a setter rather than the constructor, and deliberately so: the
     * runner is {@code EventExecutionService}, which already depends on this service for
     * {@code forceTimeEnd}. Constructor injection either way would be a cycle; the setter
     * closes the loop once, at wiring time. Null is a legal state — every test that builds
     * this service directly then behaves exactly as before Step 33.</p>
     */
    private LocationEntryPort automaticEventRunner;

    public TimeAdvancementService(TurnCycleStorePort store,
                                  UserAccessPort userAccessPort,
                                  DomainEventPublisher eventPublisher,
                                  TimeStartRecoveryService recoveryService) {
        this(store, userAccessPort, eventPublisher, recoveryService, null);
    }

    /** Step 27 — overload wiring the weather selection engine (may be null in tests). */
    public TimeAdvancementService(TurnCycleStorePort store,
                                  UserAccessPort userAccessPort,
                                  DomainEventPublisher eventPublisher,
                                  TimeStartRecoveryService recoveryService,
                                  WeatherSelectionService weatherService) {
        this.store = store;
        this.userAccessPort = userAccessPort;
        this.eventPublisher = eventPublisher;
        this.recoveryService = recoveryService;
        this.weatherService = weatherService;
    }

    /** Step 33 — see {@link #automaticEventRunner}. Called once, from the bean wiring. */
    public void setAutomaticEventRunner(LocationEntryPort automaticEventRunner) {
        this.automaticEventRunner = automaticEventRunner;
    }

    @Override
    public SleepResult sleep(String matchUuid, String userUuid) {
        long userId = requireUser(userUuid);
        MatchView match = requireMatch(matchUuid);

        // The caller must own a character in this match (mask as not-found otherwise).
        CharacterTurnView caller = store.findCharacterByMatchAndUser(match.id(), userId)
                .orElseThrow(() -> notFound());

        if (!MatchStatuses.RUNNING.equals(match.status())) {
            throw new TurnCycleException(TurnCycleException.Code.MATCH_NOT_RUNNING,
                    "Match is not RUNNING");
        }

        // Idempotent: setting sleeping on an already-sleeping character is a no-op effect.
        store.setCharacterSleeping(match.id(), caller.id(), true);
        // Step 28.7 — log the sleep action for the match logs timeline.
        store.logSleep(match.id(), caller.id(), match.currentClock());

        // Re-read the characters so the trigger sees the just-applied sleep flag.
        List<CharacterTurnView> characters = store.findCharactersByMatchId(match.id());
        boolean triggered = allCharactersDone(characters);

        int currentClock = match.currentClock();
        List<RecoveryItem> recovery = List.of();
        List<CounterZeroItem> counterZero = List.of();
        EdgeStateOutcome edgeState = EdgeStateOutcome.none();
        if (triggered) {
            AdvanceResult advanced = advanceTime(match);
            currentClock = advanced.newClock();
            recovery = advanced.recovery();
            // Step 33 — the same events, told to THIS player. The caller is the only
            // recipient with an open request; the rest learn about it over the WebSocket
            // once Steps 49-54 land, through this very method called once per player.
            counterZero = describeCounterZero(match.id(), caller.id(),
                    advanced.automaticEvents(), currentClock);
            // v0.35.6 — a time-start kills too: the recovery itself can empty a life bar, and
            // so can the events it sets off. Without this the sleeper woke up comatose with
            // nothing on screen to say why.
            edgeState = advanced.edgeState();
        }

        // After a time-end every character is awake again; otherwise the caller stays asleep.
        boolean finalSleeping = !triggered;
        return new SleepResult(matchUuid, caller.uuid(), finalSleeping, triggered, currentClock,
                recovery, counterZero, edgeState);
    }

    @Override
    public ClockResult clock(String matchUuid, String userUuid) {
        long userId = requireUser(userUuid);
        MatchView match = requireOwnedMatch(matchUuid, userId);
        return buildClock(matchUuid, match);
    }

    @Override
    public ClockResult clockForAdmin(String matchUuid) {
        MatchView match = requireMatch(matchUuid);
        return buildClock(matchUuid, match);
    }

    /** Build the clock payload (labels + per-character sleeping/energy) for a match. */
    private ClockResult buildClock(String matchUuid, MatchView match) {
        List<CharacterTurnView> characters = store.findCharactersByMatchId(match.id());
        ClockLabels labels = store.findStoryClockLabels(match.id(), DEFAULT_LANG);

        List<ClockCharacter> chars = new ArrayList<>();
        boolean anySleeping = false;
        for (CharacterTurnView c : characters) {
            anySleeping = anySleeping || c.isSleeping();
            chars.add(new ClockCharacter(c.uuid(), c.isSleeping(), c.energy()));
        }
        return new ClockResult(matchUuid, match.currentClock(),
                labels == null ? null : labels.singular(),
                labels == null ? null : labels.plural(),
                anySleeping, chars);
    }

    // ── time-end ────────────────────────────────────────────────────────────

    /**
     * Step 29 — force a time-end: put every character to sleep, then advance.
     *
     * <p>Called by {@code EventExecutionService} when an executed event carries
     * {@code flag_end_time}. It is exposed on the class and deliberately NOT on
     * {@link TimeAdvancementPort}: nothing over REST should be able to skip a time unit,
     * only the engine.</p>
     *
     * <p>Note that {@link #advanceTime} wakes everybody right after, so the net observable
     * state is "awake at clock+1". The {@code forcedSleep} flag the event returns records
     * the transition, exactly like {@code SleepResult.isSleeping = !triggered} does.</p>
     */
    public TimeEndOutcome forceTimeEnd(String matchUuid) {
        return forceTimeEnd(matchUuid, null);
    }

    /**
     * Step 33 — the same, told to a specific recipient. The event that carried
     * {@code flag_end_time} was executed by somebody, and that somebody is the one player
     * with an open request when the clock moves, so they are the one who gets to hear what
     * the time-start set off.
     */
    public TimeEndOutcome forceTimeEnd(String matchUuid, Long idRecipientCharacter) {
        MatchView match = requireMatch(matchUuid);
        store.setAllCharactersSleeping(match.id());
        AdvanceResult advanced = advanceTime(match);
        return new TimeEndOutcome(advanced.newClock(), advanced.recovery(), advanced.edgeState(),
                describeCounterZero(match.id(), idRecipientCharacter,
                        advanced.automaticEvents(), advanced.newClock()));
    }

    /**
     * Apply the per-recipient fog-of-war rule to an already-run list of automatic events.
     * Null runner (a unit test that built this service directly) means no automatic events
     * ran in the first place, so the list is empty either way.
     */
    private List<CounterZeroItem> describeCounterZero(long idMatch, Long idRecipientCharacter,
                                                      List<LocationEntryPort.AutomaticEventFired> fired,
                                                      int clock) {
        if (automaticEventRunner == null || fired == null || fired.isEmpty()) {
            return List.of();
        }
        return automaticEventRunner.describeForRecipient(idMatch, idRecipientCharacter, clock,
                fired, DEFAULT_LANG);
    }

    /** Outcome of {@link #forceTimeEnd(String)}. */
    public record TimeEndOutcome(int newClock,
                                 List<RecoveryItem> recovery,
                                 /** v0.35.6 — the edges the time-start pushed anyone over. */
                                 EdgeStateOutcome edgeState,
                                 List<CounterZeroItem> counterZero) {

        /** A time end that moved no edge. */
        public TimeEndOutcome(int newClock, List<RecoveryItem> recovery,
                              List<CounterZeroItem> counterZero) {
            this(newClock, recovery, EdgeStateOutcome.none(), counterZero);
        }
    }

    /**
     * Time-end trigger: fires when every character is sleeping OR out of energy.
     * In single-player the list is size 1. An empty list never triggers.
     */
    static boolean allCharactersDone(List<CharacterTurnView> characters) {
        if (characters == null || characters.isEmpty()) {
            return false;
        }
        for (CharacterTurnView c : characters) {
            boolean done = c.isSleeping() || c.energy() <= 0;
            if (!done) {
                return false;
            }
        }
        return true;
    }

    private AdvanceResult advanceTime(MatchView match) {
        int newClock = store.incrementMatchClock(match.id());
        store.insertClockHistory(match.id(), newClock);
        store.wakeAllCharacters(match.id());
        // Step 26: per-character recovery, class bonuses and location counters.
        TimeStartRecoveryService.TimeStartOutcome outcome =
                recoveryService.applyAtTimeStart(match.id());
        // Step 33: the events that pass collected — counters that reached zero, and the
        // locations whose id_event_if_character_start_time fires because a time unit began
        // with somebody standing there. Run here rather than inside the recovery service:
        // the event engine sits above it in the wiring, and an event can force a time end.
        List<LocationEntryPort.AutomaticEventFired> fired = automaticEventRunner == null
                ? List.of()
                : automaticEventRunner.runPendingAutomaticEvents(
                        match.id(), newClock, outcome.pending(), DEFAULT_LANG);
        // Step 27: select the weather for the new time unit and apply its energy delta.
        if (weatherService != null) {
            weatherService.applyAtTimeStart(match.id());
        }
        rebuildQueue(match.id(), newClock);
        eventPublisher.publish(new TimeAdvanced(match.uuid(), newClock));
        List<RecoveryItem> recovery = new ArrayList<>();
        for (TimeStartRecoveryService.RecoveryRecap r : outcome.recovery()) {
            recovery.add(new RecoveryItem(r.characterUuid(), r.energyDelta(), r.lifeDelta(), r.sadDelta()));
        }
        // The recovery's own verdict first, then whatever its events did: one edge state.
        List<EdgeStateOutcome> parts = new ArrayList<>();
        parts.add(outcome.edgeState());
        for (LocationEntryPort.AutomaticEventFired f : fired) {
            parts.add(f.edgeState());
        }
        return new AdvanceResult(newClock, recovery, fired, EdgeStateOutcome.merge(parts));
    }

    private record AdvanceResult(int newClock,
                                 List<RecoveryItem> recovery,
                                 List<LocationEntryPort.AutomaticEventFired> automaticEvents,
                                 EdgeStateOutcome edgeState) {
    }

    /** Rebuild the turn queue for a new clock: all WAITING, highest priority ACTIVE. */
    private void rebuildQueue(long idMatch, int clock) {
        List<CharacterTurnView> characters = store.findCharactersByMatchId(idMatch);
        if (characters.isEmpty()) {
            store.replaceQueue(idMatch, List.of());
            store.updateMatchStatusAndTurn(idMatch, MatchStatuses.RUNNING, null);
            return;
        }
        List<QueueRow> rows = new ArrayList<>();
        for (CharacterTurnView c : characters) {
            long priority = TurnPriorityCalculator.compute(
                    c.dexterity(), c.intelligence(), c.constitution(), c.life(), c.id());
            rows.add(new QueueRow(c.id(), null, clock, priority, TurnStatuses.WAITING, 0, null, null));
        }
        rows.sort(Comparator.comparingLong(QueueRow::priority).reversed());

        QueueRow top = rows.get(0);
        rows.set(0, new QueueRow(top.idCharacterMatch(), top.uuid(), top.clock(), top.priority(),
                TurnStatuses.ACTIVE, top.passCounter(), null, null));

        store.replaceQueue(idMatch, rows);
        store.updateMatchStatusAndTurn(idMatch, MatchStatuses.RUNNING, top.idCharacterMatch());
    }

    // ── helpers ───────────────────────────────────────────────────────────

    private long requireUser(String userUuid) {
        return userAccessPort.findByUuid(userUuid)
                .map(UserAccessPort.UserView::id)
                .orElseThrow(() -> notFound());
    }

    private MatchView requireMatch(String matchUuid) {
        return store.findMatchByUuid(matchUuid).orElseThrow(() -> notFound());
    }

    private MatchView requireOwnedMatch(String matchUuid, long userId) {
        MatchView match = requireMatch(matchUuid);
        if (match.idUserCreator() != userId) {
            throw notFound();
        }
        return match;
    }

    private static TurnCycleException notFound() {
        return new TurnCycleException(TurnCycleException.Code.MATCH_NOT_FOUND,
                "Match not found or not accessible");
    }
}

File: code/backend/python/app/core/services/match/time_advancement_service.py
"""Step 25 — time advancement & clock cycle service (mirrors the Java reference).

Adds a voluntary sleep action and a clock query, and advances the match clock
when every active character is "done" (sleeping or out of energy — in
single-player the list is size 1). On time-end it logs the advance, wakes the
characters, rebuilds ``gaming_turn_queue`` reusing the Step 24 priority formula,
and publishes a ``TimeAdvanced`` event.
"""
from typing import Any, Dict, List

from app.core.models.match import match_statuses
from app.core.models.match import turn_models as tm
from app.core.models.match.time_models import (
    ClockCharacter,
    ClockResult,
    RecoveryItem,
    SleepResult,
    TimeAdvanced,
)
from app.core.models.match.turn_models import TurnCycleError
from app.core.ports.event.event_ports import DomainEventPublisher
from app.core.ports.match.time_ports import TimeAdvancementPort, TimeStorePort
from app.core.services.match.time_start_recovery_service import TimeStartRecoveryService

DEFAULT_LANG = "en"


class TimeAdvancementService(TimeAdvancementPort):
    def __init__(self, store: TimeStorePort, event_publisher: DomainEventPublisher,
                 recovery_service: "TimeStartRecoveryService | None" = None,
                 weather_service=None, edge_store=None) -> None:
        self.store = store
        self.event_publisher = event_publisher
        self.recovery_service = recovery_service or TimeStartRecoveryService(store, edge_store)
        # Step 27 — optional weather selection engine (may be None in tests).
        self.weather_service = weather_service
        # Step 33 — the engine that runs the automatic events a time-start collected.
        # Injected through a setter rather than the constructor, and deliberately so: the
        # runner is EventService, which already depends on this service for
        # force_time_end. Constructor injection either way would be a cycle; the setter
        # closes the loop once, at wiring time. None is a legal state — every test that
        # builds this service directly then behaves exactly as before Step 33.
        self.automatic_event_runner = None

    def set_automatic_event_runner(self, runner) -> None:
        """Step 33 — see ``automatic_event_runner``. Called once, from the wiring."""
        self.automatic_event_runner = runner

    # ── public API ──────────────────────────────────────────────────────────

    def sleep(self, match_uuid: str, user_uuid: str) -> SleepResult:
        user_id = self._require_user(user_uuid)
        match = self._require_match(match_uuid)

        # The caller must own a character in this match (mask as not-found otherwise).
        caller = self.store.find_character_by_match_and_user(match["id"], user_id)
        if caller is None:
            raise TurnCycleError(TurnCycleError.MATCH_NOT_FOUND, "Match not found or not accessible")

        if match["status"] != match_statuses.RUNNING:
            raise TurnCycleError(TurnCycleError.MATCH_NOT_RUNNING, "Match is not RUNNING")

        # Idempotent: setting sleeping on an already-sleeping character is a no-op effect.
        self.store.set_character_sleeping(match["id"], caller["id"], True)
        # Step 28.7 — log the sleep action for the match logs timeline.
        self.store.log_sleep(match["id"], caller["id"], match.get("current_clock") or 0)

        # Re-read so the trigger sees the just-applied sleep flag.
        characters = self.store.find_characters_by_match_id(match["id"]) or []
        triggered = self._all_done(characters)

        current_clock = match["current_clock"]
        recovery: List[RecoveryItem] = []
        counter_zero: List[Any] = []
        if triggered:
            current_clock, recovery, fired = self._advance_time(match)
            # Step 33 — the same events, told to THIS player. The caller is the only
            # recipient with an open request; the rest learn about it over the WebSocket
            # once Steps 49-54 land, through this very method called once per player.
            counter_zero = self._describe_counter_zero(match["id"], caller["id"], fired,
                                                       current_clock)

        # After a time-end every character is awake again; otherwise the caller stays asleep.
        return SleepResult(match_uuid, caller["uuid"], not triggered, triggered,
                           current_clock, recovery, counter_zero)

    def clock(self, match_uuid: str, user_uuid: str) -> ClockResult:
        user_id = self._require_user(user_uuid)
        match = self._require_owned_match(match_uuid, user_id)

        characters = self.store.find_characters_by_match_id(match["id"]) or []
        singular, plural = self.store.find_story_clock_labels(match["id"], DEFAULT_LANG)

        chars: List[ClockCharacter] = []
        any_sleeping = False
        for c in characters:
            sleeping = bool(c.get("is_sleeping"))
            any_sleeping = any_sleeping or sleeping
            chars.append(ClockCharacter(c["uuid"], sleeping, c.get("energy", 0) or 0))
        return ClockResult(match_uuid, match["current_clock"], singular, plural,
                           any_sleeping, chars)

    # ── time-end ──────────────────────────────────────────────────────────────

    @staticmethod
    def _all_done(characters: List[Dict[str, Any]]) -> bool:
        """Time-end trigger: every character is sleeping OR out of energy.
        In single-player the list is size 1. An empty list never triggers."""
        if not characters:
            return False
        for c in characters:
            done = bool(c.get("is_sleeping")) or (c.get("energy", 0) or 0) <= 0
            if not done:
                return False
        return True

    def force_time_end(self, match_uuid: str) -> int:
        """Step 29 — force a time end: put every character to sleep, then advance.

        Exposed on the class and deliberately NOT on TimeAdvancementPort: nothing over REST
        should be able to skip a time unit, only the engine. Note that _advance_time wakes
        everybody right after, so the net observable state is "awake at clock+1"; the
        forced_sleep flag the event returns records the transition.
        """
        match = self._require_match(match_uuid)
        self.store.set_all_characters_sleeping(match["id"])
        new_clock, _recovery, _fired = self._advance_time(match)
        return new_clock

    def _advance_time(self, match: Dict[str, Any]):
        new_clock = self.store.increment_match_clock(match["id"])
        self.store.insert_clock_history(match["id"], new_clock)
        self.store.wake_all_characters(match["id"])
        # Step 26: per-character recovery, class bonuses and location counters.
        outcome = self.recovery_service.apply_at_time_start(match["id"])
        # Step 33: the events that pass collected — counters that reached zero, and the
        # locations whose id_event_if_character_start_time fires because a time unit began
        # with somebody standing there. Run here rather than inside the recovery service:
        # the event engine sits above it in the wiring, and an event can force a time end.
        fired: List[Any] = []
        if self.automatic_event_runner is not None and outcome.pending:
            fired = self.automatic_event_runner.run_pending_automatic_events(
                match["id"], new_clock, outcome.pending, DEFAULT_LANG)
        # Step 27: select the weather for the new time unit and apply its delta.
        if self.weather_service is not None:
            self.weather_service.apply_at_time_start(match["id"])
        self._rebuild_queue(match["id"], new_clock)
        self.event_publisher.publish(TimeAdvanced(match["uuid"], new_clock))
        return new_clock, outcome.recovery, fired

    def _describe_counter_zero(self, id_match: int, id_recipient_character,
                               fired: List[Any], clock: int) -> List[Any]:
        """Apply the per-recipient fog-of-war rule to an already-run list.

        No runner (a unit test that built this service directly) means no automatic events
        ran in the first place, so the list is empty either way.
        """
        if self.automatic_event_runner is None or not fired:
            return []
        return self.automatic_event_runner.describe_for_recipient(
            id_match, id_recipient_character, clock, fired, DEFAULT_LANG)

    def _rebuild_queue(self, id_match: int, clock: int) -> None:
        characters = self.store.find_characters_by_match_id(id_match) or []
        if not characters:
            self.store.replace_queue(id_match, [])
            self.store.update_match_status_and_turn(id_match, match_statuses.RUNNING, None)
            return
        rows: List[Dict[str, Any]] = []
        for c in characters:
            priority = tm.compute_priority(c["dexterity"], c["intelligence"],
                                           c["constitution"], c["life"], c["id"])
            rows.append({
                "id_character_match": c["id"], "clock": clock, "priority": priority,
                "status": tm.WAITING, "pass_counter": 0,
                "timestamp_start": None, "timestamp_end": None,
            })
        rows.sort(key=lambda r: r["priority"], reverse=True)
        rows[0]["status"] = tm.ACTIVE
        self.store.replace_queue(id_match, rows)
        self.store.update_match_status_and_turn(id_match, match_statuses.RUNNING,
                                                rows[0]["id_character_match"])

    # ── helpers ───────────────────────────────────────────────────────────────

    def _require_user(self, user_uuid: str) -> int:
        user_id = self.store.find_user_id_by_uuid(user_uuid)
        if user_id is None:
            raise TurnCycleError(TurnCycleError.MATCH_NOT_FOUND, "Match not found or not accessible")
        return user_id

    def _require_match(self, match_uuid: str) -> Dict[str, Any]:
        match = self.store.find_match_by_uuid(match_uuid)
        if match is None:
            raise TurnCycleError(TurnCycleError.MATCH_NOT_FOUND, "Match not found or not accessible")
        return match

    def _require_owned_match(self, match_uuid: str, user_id: int) -> Dict[str, Any]:
        match = self._require_match(match_uuid)
        if match["id_user_creator"] != user_id:
            raise TurnCycleError(TurnCycleError.MATCH_NOT_FOUND, "Match not found or not accessible")
        return match
